File: src/test/suite/lineRanges.test.ts
import * as assert from "assert";
import { collectLineRanges, formatLineRanges, selectionToLineRange } from "../../core/lineRanges";

suite("lineRanges", () => {
  test("selectionToLineRange returns current line for empty selection", () => {
    const result = selectionToLineRange({
      startLine: 7,
      startCharacter: 3,
      endLine: 7,
      endCharacter: 3,
      isEmpty: true
    });

    assert.deepStrictEqual(result, { start: 8, end: 8 });
  });

  test("selectionToLineRange excludes trailing line when selection ends at column zero", () => {
    const result = selectionToLineRange({
      startLine: 0,
      startCharacter: 0,
      endLine: 3,
      endCharacter: 0,
      isEmpty: false
    });

    assert.deepStrictEqual(result, { start: 1, end: 3 });
  });

  test("collectLineRanges sorts and merges overlapping or adjacent ranges", () => {
    const result = collectLineRanges([
      {
        startLine: 7,
        startCharacter: 0,
        endLine: 10,
        endCharacter: 4,
        isEmpty: false
      },
      {
        startLine: 0,
        startCharacter: 0,
        endLine: 2,
        endCharacter: 5,
        isEmpty: false
      },
      {
        startLine: 4,
        startCharacter: 1,
        endLine: 5,
        endCharacter: 0,
        isEmpty: false
      },
      {
        startLine: 2,
        startCharacter: 8,
        endLine: 5,
        endCharacter: 1,
        isEmpty: false
      }
    ]);

    assert.deepStrictEqual(result, [
      { start: 1, end: 6 },
      { start: 8, end: 11 }
    ]);
  });

  test("formatLineRanges matches sparse multi-selection output", () => {
    const result = formatLineRanges(
      collectLineRanges([
        {
          startLine: 0,
          startCharacter: 0,
          endLine: 2,
          endCharacter: 2,
          isEmpty: false
        },
        {
          startLine: 4,
          startCharacter: 0,
          endLine: 5,
          endCharacter: 3,
          isEmpty: false
        },
        {
          startLine: 7,
          startCharacter: 0,
          endLine: 10,
          endCharacter: 1,
          isEmpty: false
        }
      ])
    );

    assert.strictEqual(result, "1-3,5-6,8-11");
  });

  test("formatLineRanges renders discrete and continuous ranges", () => {
    const result = formatLineRanges([
      { start: 1, end: 3 },
      { start: 5, end: 6 },
      { start: 8, end: 11 }
    ]);

    assert.strictEqual(result, "1-3,5-6,8-11");
  });
});

File: idea-plugin/src/test/kotlin/com/dmxiaoshubao/copypathline/idea/core/LineRangesTest.kt
package com.dmxiaoshubao.copypathline.idea.core

import kotlin.test.Test
import kotlin.test.assertEquals

class LineRangesTest {
    @Test
    fun `empty selection uses current line`() {
        val result = LineRanges.collect(
            text = sampleText(),
            selections = listOf(SelectionRange(startOffset = 2, endOffset = 2, hasSelection = false)),
        )

        assertEquals(listOf(LineRange(1, 1)), result)
    }

    @Test
    fun `selection ending at next line start excludes trailing line`() {
        val text = "one\ntwo\nthree\n"
        val result = LineRanges.collect(
            text = text,
            selections = listOf(
                SelectionRange(
                    startOffset = 0,
                    endOffset = text.indexOf("three"),
                    hasSelection = true,
                ),
            ),
        )

        assertEquals(listOf(LineRange(1, 2)), result)
    }

    @Test
    fun `ranges are sorted and merged`() {
        val text = sampleText()
        val result = LineRanges.collect(
            text = text,
            selections = listOf(
                SelectionRange(text.indexOf("eight"), text.length, true),
                SelectionRange(0, text.indexOf("three"), true),
                SelectionRange(text.indexOf("three"), text.indexOf("six"), true),
                SelectionRange(text.indexOf("five"), text.indexOf("six"), true),
            ),
        )

        assertEquals(listOf(LineRange(1, 5), LineRange(8, 9)), result)
    }

    @Test
    fun `format matches sparse multi selection output`() {
        val text = sampleText()
        val ranges = LineRanges.collect(
            text = text,
            selections = listOf(
                SelectionRange(0, text.indexOf("four"), true),
                SelectionRange(text.indexOf("five"), text.indexOf("seven"), true),
                SelectionRange(text.indexOf("eight"), text.length, true),
            ),
        )

        assertEquals("1-3,5-6,8-9", LineRanges.format(ranges))
    }

    @Test
    fun `crlf text is handled correctly`() {
        val text = "a\r\nb\r\nc\r\n"
        val result = LineRanges.collect(
            text = text,
            selections = listOf(SelectionRange(0, text.indexOf("c"), true)),
        )

        assertEquals(listOf(LineRange(1, 2)), result)
    }

    private fun sampleText(): String = """
        one
        two
        three
        four
        five
        six
        seven
        eight
        nine
    """.trimIndent()
}
